### PR TITLE
Cancel and remove export

### PR DESCRIPTION
**Pull request by forus**

### CODE REVIEW

File: open-api/swagger.yaml [hunk 1]
```diff
@@ -1783,6 +1783,22 @@ paths:
...
+      parameters:
+        - name: jobId
+          required: true
+          in: path
+          description: 'Id of the export job.'
+          type: string
+      responses:
+        200:
+          description: |
+            200 http status whne job has been cancelled.
```
Comment (gijskant) [inline open-api/swagger.yaml]: typo

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy [hunk 2]
```diff
@@ -68,19 +69,39 @@ class ExportAsyncJobService {
...
+                    " Setting the job flag as cancelled anyway.")
+        }
+        updateStatus(jobId, JobStatus.CANCELLED)
+    }
+
+    /**
+     * Deletes a job after the cancelling it
+     */
+    void deleteJob(Long jobId) {
+        cancelJob(jobId)
```
Comment (gijskant) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy]: Or check status to be `CANCELLED` or `FINISHED`?

**Comment by gijskant** [comment]
Flow diagram:
```
                                +-----------+
             +------------------>   Error   <-----+
             |                  +-^---------+     |
     +       |                    |     |         |
     |       |                    |     |         |
     |       |                    |     |         |
  +--v----------+     +-------------+   | +----------------+     +---------------+
  |   Created   +----->   Started   +-----> Gathering data +----->   Completed   |
  +-------------+     +-------------+   | +----------------+     +---------------+
             |                    |     |  |                         |
             |                    |     |  |                         |
             |                    |     |  |                         |
             |                  +-v-----v--v+                        |
             +------------------> Cancelled <------------------------+
                                +-----------+
```

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy [hunk 3]
```diff
@@ -68,19 +69,39 @@ class ExportAsyncJobService {
...
+    void cancelJob(Long jobId) {
+        AsyncJobCoreDb job = AsyncJobCoreDb.findById(jobId)
+        if (!job) {
+            String errMsg = "Job with id ${jobId} hasn't been found."
+            log.error(errMsg)
+            throw new NoSuchResourceException("Job with id ${jobId} hasn't been found.")
+        }
+
+        if (isJobCancelled(job)) {
+            log.info("Job with ${jobId} id has been canelled already.")
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy]: typo

**Comment by ewelinagr** [comment]
@forus In master branch there is a change with adding unique identity for each new trigger, maybe it's useful to have it here as well? https://github.com/thehyve/transmart-core/commit/2d83969c8c3a1626fa411856f7e880d015f35cb4

**Comment by forus** [comment]
Thanks! All feedback was useful, I've applied all suggestions.

@gijskant We throw an exception when we try to cancel a job with `Error` or `Complete` status see `inTerminalStatus`. There is also a separate audit log interceptor for export.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy [hunk 4]
```diff
@@ -68,44 +73,65 @@ class ExportAsyncJobService {
...
-        jsonResult.put("jobId", id)
-        return jsonResult
+    void deleteJob(Long jobId) {
+        AsyncJobCoreDb job = getJobById(jobId)
+        if (!job.inTerminalStatus) {
+            log.info("Job with id ${jobId} hasn't finished. Try to cancel it first.")
+            try {
+                cancelJob(jobId)
+            } catch (Exception e) {
+                log.error("Can't cancel jop with id ${jobId}.", e)
```
Comment (gijskant) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy]: Typo.

File: transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy [hunk 5]
```diff
@@ -14,49 +16,45 @@ import java.util.zip.ZipOutputStream
...
     void execute(JobExecutionContext jobExecutionContext) {
         Map jobDataMap = jobExecutionContext.jobDetail.getJobDataMap()
 
-        PersistenceContextInterceptor interceptor
+        PersistenceContextInterceptor interceptor = ctx.persistenceInterceptor
         try {
-            interceptor = Holders.applicationContext.persistenceInterceptor
             interceptor.init()
-            zipData(jobDataMap)
-        } catch (UnexpectedResultException e) {
```
Comment (gijskant) [inline transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy]: I agree that it's much cleaner like this, but now too many possible cases are missed this way. Consider the case that some filesystem error occurs when accessing the working directory. Also, please change `Exception` to `Throwable`, now assertion errors are not caught).

**Pull request merged**